src/training/graph_group.h

```diff
@@ -112,12 +112,12 @@ class AsyncGraphGroup : public GraphGroup {
           graph->forward();
         }
 
-        if(shardOpt_.size() == 0){
-          for (auto device : devices_) {
+        if(shardOpt_.size() == 0)
+          for (auto device : devices_) 
             shardOpt_.push_back(Optimizer(options_));
 
-          }
-        }
+        shardSync_.reset( new std::mutex[devices_.size()] );
+         
```

Comment:
This line is actually highly worrying. You destroy the mutexes for each batch, probably while other threads are still locked at them. The interesting part though is, when I fix this, convergence becomes worse.